v0/destinations/singular/transform.js

```diff
@@ -0,0 +1,215 @@
+  if (
+    Config.sessionEventList.includes(eventType) ||
+    sessionEvents.includes(eventType)
+  ) {
+    if (platform.toLowerCase() === "android") {
+      payload = constructPayload(
+        message,
+        MAPPING_CONFIG[CONFIG_CATEGORIES.SESSION_ANDROID.name]
+      );
+      if (!payload) {
+        // fail-safety for developer error
+        throw new CustomError("Failed to Create Android Session Payload", 400);
+      }
+      extractCustomFields(
+        message,
+        eventAttributes,
+        ["properties"],
+        SINGULAR_SESSION_ANDROID_EXCLUSION
+      );
+    } else if (platform.toLowerCase() === "ios") {
+      payload = constructPayload(
+        message,
+        MAPPING_CONFIG[CONFIG_CATEGORIES.SESSION_IOS.name]
+      );
+      if (!payload) {
+        // fail-safety for developer error
+        throw new CustomError("Failed to Create iOS Session Payload", 400);
+      }
+      extractCustomFields(
+        message,
+        eventAttributes,
+        ["properties"],
+        SINGULAR_SESSION_IOS_EXCLUSION
+      );
+
+      /*
+        if att_authorization_status is true then dnt will be false,
+        else by default dnt value is true
+      */
+      payload.dnt = !payload.att_authorization_status;
+    } else {
+      throw new CustomError("[Singular] :: Invalid Platform", 400);
+    }
+
+    // Singular maps Connection Type to either wifi or carrier
+    if (message.context?.network?.wifi === "wifi") {
+      payload.c = "wifi";
+    } else {
+      payload.c = "carrier";
+    }
+
+    payload = removeUndefinedAndNullAndEmptyValues(payload);
+    endPoint = `${BASE_URL}/launch`;
+  } else {
+    // event payload
+    if (platform.toLowerCase() === "android") {
+      payload = constructPayload(
+        message,
+        MAPPING_CONFIG[CONFIG_CATEGORIES.EVENT_ANDROID.name]
+      );
+      if (!payload) {
+        // fail-safety for developer error
+        throw new CustomError("Failed to Create Android Event Payload", 400);
+      }
+      extractCustomFields(
+        message,
+        eventAttributes,
+        ["properties"],
+        SINGULAR_EVENT_ANDROID_EXCLUSION
+      );
+    } else if (platform.toLowerCase() === "ios") {
+      payload = constructPayload(
+        message,
+        MAPPING_CONFIG[CONFIG_CATEGORIES.EVENT_IOS.name]
+      );
+      if (!payload) {
+        // fail-safety for developer error
+        throw new CustomError("Failed to Create iOS Event Payload", 400);
+      }
+      extractCustomFields(
+        message,
+        eventAttributes,
+        ["properties"],
+        SINGULAR_EVENT_IOS_EXCLUSION
+      );
+    } else {
+      throw new CustomError("[Singular] :: Invalid Platform", 400);
+    }
+    const productList = message.properties?.products;
+    if (productList && Array.isArray(productList)) {
+      const responseArray = [];
+      const finalPayload = payload;
+      for (const product of productList) {
+        const productDetails = constructPayload(
```

Comment:
It can be possible the `productDetatils` can be an empty payload. Please provide a check.

Reply:
https://github.com/rudderlabs/rudder-transformer/blob/0ca3cc019e8aa26c54f7928671225f6e5766b063/v0/destinations/revenue_cat/transform.js#L32
I have taken these suggestions from the RevenueCat implementations, should I put check conditions here?

Reply:
Okay let it be in that way. We have to think regarding this as we can't throw error for a single product and skipping silently is also not the answer.